### PR TITLE
fix(helm): Wait for CRDs to reach established state for crd_install hook

### LINKED ISSUE #4925: crd-install hook possible race condition
While installing the `stable/prometheus-operator` chart, which creates 4 CRDs using the `crd-install` hooks, it is possible for the CRDs to be created but the validation fail anyway with 

```
Error: validation failed: [unable to recognize "": no matches for kind "Alertmanager" in version
"monitoring.coreos.com/v1", unable to recognize "": no matches for kind "Prometheus" in version 
"monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version 
. . .
```

After seeing this error, repeatedly hitting `kubectl get crd | grep coreos` will show that at first there are <4 CRDs but eventually all 4 are created. 

This behaviour is absolutely consistent on a specific cluster 
- if it occurs it will continue to behave this way in the cluster 
- if it does not occur then it cannot be reproduced on that cluster 

There are multiple reports of this happening with this chart but in most cases there are no issues, including the helm/charts CI process. 

Issue on helm/charts: https://github.com/helm/charts/issues/9241

---

#### Can be broken 
Confirmed:
```
helm:
Client: &version.Version{SemVer:"v2.10.0", GitCommit:"9ad53aac42165a5fadc6c87be0dea6b115f93090", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.10.0", GitCommit:"9ad53aac42165a5fadc6c87be0dea6b115f93090", GitTreeState:"clean"}

kubectl:
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.1", GitCommit:"4ed3216f3ec431b140b1d899130a69fc671678f4", GitTreeState:"clean", BuildDate:"2018-10-05T16:46:06Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.4", GitCommit:"bf9a868e8ea3d3a8fa53cbb22f566771b3f8068b", GitTreeState:"clean", BuildDate:"2018-10-25T19:06:30Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```
Reported:
helm: 2.10, 2.11
kubectl: 1.11.2, 1.11.3, 1.11.4

#### Can work 
Confirmed:
```
helm:
Client: &version.Version{SemVer:"v2.10.0", GitCommit:"9ad53aac42165a5fadc6c87be0dea6b115f93090", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.10.0", GitCommit:"9ad53aac42165a5fadc6c87be0dea6b115f93090", GitTreeState:"clean"}

kubectl:
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.1", GitCommit:"4ed3216f3ec431b140b1d899130a69fc671678f4", GitTreeState:"clean", BuildDate:"2018-10-05T16:46:06Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.2", GitCommit:"81753b10df112992bf51bbc2c2f85208aad78335", GitTreeState:"clean", BuildDate:"2018-04-27T09:10:24Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```
```
helm:
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}

kubectl:
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.1", GitCommit:"4ed3216f3ec431b140b1d899130a69fc671678f4", GitTreeState:"clean", BuildDate:"2018-10-05T16:46:06Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.1", GitCommit:"b1b29978270dc22fecc592ac55d903350454310a", GitTreeState:"clean", BuildDate:"2018-07-17T18:43:26Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```

#### Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
- Azure 
- possibly others

### PR DESCRIPTION
**Pull request by mortent**
**What this PR does / why we need it**:
There is a race condition in the crd_install hook implementation, where there is a chance a CRD is not yet ready by the time CRs are being created. This is reported in issue #4925. This change makes sure CRDs installed through the crd_install hook reaches the `established` state before the hook is considered complete.

Fixes #4925

**Special notes for your reviewer**:
Unit-testing code in the kubernetes client is difficult, as the builder/infos is tightly coupled with the API server. I will look into how to improve testing for this part of the codebase, but I would like to separate it from this PR.

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by mortent** [comment]
Updated this PR to include unit tests.

**Comment by aviau** [comment]
I am curious if this also fixes race conditions in chart purges?

**Review by bacongobbler — APPROVED** [review]
the code in here looks great to me!

**Comment by vsliouniaev** [comment]
Two more issues opened that should get fixed by this PR
https://github.com/coreos/prometheus-operator/issues/2432
https://github.com/helm/charts/issues/11702

**Comment by micw** [comment]
I tested this with the test chart+script I wrote to reproduce #5375 - the issue still occurs.

**Comment by jhohertz** [comment]
CRDs with helm, and in particular trying to deploy prometheus-operator chart is completely broken until something like this gets merged down.

**Comment by tim-beatson** [comment]
@bacongobbler, I'm pretty surprised this didn't make it into Helm 2.13.0. Are you able to offer a timescale on when we can see this in a release of Helm? I have a number of charts I want to use that are completely broken until this can be resolved....

**Comment by Starefossen** [comment]
What's the holdup of this? Anything we can do to help moving this forward?

**Comment by bacongobbler** [comment]
Any PR size/L or higher needs two maintainers to sign off. I'll see if I can but one of the other maintainers to take a look at this before the 2.14 release. Sorry for the delay on this one!

**Comment by tim-beatson** [comment]
@bacongobbler, that would be massively appreciated.

**Comment by helgi** [comment]
Getting the conflict resolved and this landed would be great :-)

**Comment by micw** [comment]
I'd say, the conflict is not the main problem. Please have a look at my comment above. I build and deployed tiller with this PR applied and it _did not_ solved the issue (see https://github.com/helm/helm/issues/5375).

**Comment by mortent** [comment]
I have rebased the PR so the conflict should be resolved.

@micw Your issue is actually a separate one from what is being addressed in this PR. This fixes the issue where the installed CRD is not ready at the point where CRs are created. 

Your issue is that the deletion of the CRD through the hook-delete-policy hook has not completed when the crd-install hook tries to install the same CRD. Based on the error message, it sounds like the deletion timestamp has been set, but the resource is not yet removed from etcd. From just a quick look at the code that does the deletion, I think it needs to poll for the resource to be fully removed from etcd before the delete hook should be considered complete. I will try to find time to make a separate fix for this issue.

**Comment by Morriz** [comment]
Any updates? Can this be merged?

**Comment by bacongobbler** [comment]
@Morriz see my [comment above](https://github.com/helm/helm/pull/5112#issuecomment-475929589)

**Comment by thomastaylor312** [comment]
I'll be reviewing this soon

**Comment by squillace** [comment]
@thomastaylor312 I know you're busy, but I, too, am hoping this can get reviewed sooner rather than later. MEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE.

**Comment by jdn-za** [comment]
+1 to that, anything we can help with?

**Review by thomastaylor312 — APPROVED** [review]
This is good to go! Thanks for all the patience everyone and @mortent for the fix. 

@bacongobbler Merge away!

**Comment by guilledipa** [comment]
Is there a time estimate to port this change to Helm v3?

Thanks for this!

**Comment by thomastaylor312** [comment]
@guilledipa thanks for the reminder! We missed labeling this with the needs v3 label so we can make sure to come back around and port it over

**Comment by thomastaylor312** [comment]
This was implemented in Helm 3 by #6332

**Pull request merged**
